src/ast/transform/MinimiseProgram.cpp

```diff
@@ -316,8 +316,11 @@ bool MinimiseProgramTransformer::reduceSingletonRelations(TranslationUnit& trans
             if (areBijectivelyEquivalent(normedFirst, normedSecond)) {
                 QualifiedName firstName = first->getHead()->getQualifiedName();
                 QualifiedName secondName = second->getHead()->getQualifiedName();
-                redundantClauses.insert(second);
-                canonicalName.insert(std::pair(secondName, firstName));
+                // check whether relations have same definitions
+                if (*getRelation(program, firstName) == *getRelation(program, secondName)) {
```

Comment:
As far as I know, this gets rid of the main optimisation opportunity this method is trying to find; namely, when two relations that are different (and therefore have different qualified names) have equivalent singleton clauses. 

E.g., will the following program still be optimised:

```
.decl in(X:number)
.decl a(X:number)
.decl b(Y:number)
.decl c(X:number)
in(0).
a(X) :- in(X), X != 1.
b(Y) :- Y != 1, in(Y).
c(X) :- a(X), b(X).
.output c()
```

into:

```
.decl in(X:number)
.decl a(X:number)
.decl c(X:number)
in(0).
a(X) :- in(X), X != 1.
c(X) :- a(X).
.output c()
```

Instead, I think the check should be skipped for relations tagged with inline, eqrel, etc., (like we do for input relations; see line 290 in this file) *or* just check the qualifiers for relations.

Reply:
Also, I wanted to ask whether the types of the attributes are checked in the current version.

Reply:
Not currently. My assumption was that if two clauses were found to be bijectively equivalent, then all variables should have the same type information (ignoring the head), and so it could be safely ignored. Do you think this assumption breaks?

Reply:
I think this is not quite true considering that we have a more complex type system than in the old days. 
With union-types, we could have identical rules but different relation schemas. A more fatal situation would arise related to number vs. unsigned attributes etc. 

To summarise, I feel we need to check for equivalence on each attribute type, relation representation and qualifiers.